File: python/test_sorting.py
import pytest
import sorting
from sorting import is_sorted, is_sorted_until, partial_sort, \
        partial_sort_copy, nth_element

@pytest.fixture
def ary():
    return [2, 3, 1, 4, 5]

def test_is_sorted_until(ary):
    assert is_sorted_until(ary) == 2
    sorted_ary = [1, 2, 3]
    assert is_sorted_until(sorted_ary) == len(sorted_ary)

def test_is_sorted(ary):
    assert not is_sorted(ary)
    ary.sort()
    assert is_sorted(ary)

def test_sort(ary):
    assert not is_sorted(ary)
    sorting.sort(ary)
    assert is_sorted(ary)

def test_partial_sort(ary):
    partial_sort(ary, 2)
    assert ary[0] == 1
    assert ary[1] == 2
    partial_sort(ary, 3)
    assert ary[2] == 3

def test_partial_sort_copy(ary):
    assert partial_sort_copy(ary, 2) == [1, 2]

def test_stable_sort():
    pairs = [(1, 0), (2, 1), (2, 2), (2, 3), (1, 2), (3, 1)]
    sorting.stable_sort(pairs, lambda x, y: x[0] < y[0])
    assert pairs == sorted(pairs, key=lambda x: x[0])

def test_nth_element(ary):
    nth_element(ary, 3)
    assert ary[2] == 3
    nth_element(ary, 2)
    assert ary[1] == 2

File: ruby/test_sorting.rb
require 'minitest/autorun'
require_relative 'sorting'

class TestSorting < MiniTest::Test
  include Sorting

  def setup
    @a = [2, 3, 1, 4, 5]
  end

  def test_is_sorted_until
    assert_equal 2, is_sorted_until(@a)
    sorted = [1, 2, 3]
    assert_equal sorted.size, is_sorted_until(sorted)
  end

  def test_is_sorted
    assert_equal false, is_sorted(@a)
    assert_equal true, is_sorted(@a.sort)
  end

  def test_sort
    assert_equal false, is_sorted(@a)
    sort(@a)
    assert_equal true, is_sorted(@a)
  end

  def test_partial_sort
    partial_sort(@a, 2)
    assert_equal 1, @a[0]
    assert_equal 2, @a[1]
    partial_sort(@a, 3)
    assert_equal 3, @a[2]
  end

  def test_partial_sort_copy
    assert_equal [1, 2], partial_sort_copy(@a, 2)
  end

  def test_stable_sort
    aries = [
      [1, 0], [2, 1], [2, 3], [2, 2], [1, 2], [3, 1]
    ]
    stable_sort(aries){|x, y| x[0] < y[0]}
    sorted = [
      [1, 0], [1, 2], [2, 1], [2, 3], [2, 2], [3, 1]
    ]
    assert_equal sorted, aries
  end

  def test_nth_element
    assert_equal 3, nth_element(@a, 2)
    assert_equal 1, nth_element(@a, 0)
  end
end
